src/transformStateWithClones.js

```diff
@@ -6,8 +6,43 @@
  *
  * @return {Object[]}
  */
+
 function transformStateWithClones(state, actions) {
-  // write code here
+  const states = [];
+
+  for (const action of actions) {
+    const { type = null, extraData = {}, keysToRemove = [] } = action;
+
+    let newObj;
+
+    if (states.length === 0) {
+      newObj = { ...state };
+    } else {
+      newObj = { ...states[states.length - 1] };
+    }
+
+    switch (type) {
+      case 'addProperties':
+        newObj = Object.assign(newObj, extraData);
+        break;
+      case 'removeProperties':
+        keysToRemove.forEach((element) => {
+          if (newObj[element]) {
+            delete newObj[element];
+          }
+        });
+        break;
+      case 'clear':
+        Object.keys(newObj).forEach((element) => {
+          delete newObj[element];
+        });
+        break;
+    }
+
+    states.push(newObj);
```

Comment:
The logic to create a new state object based on the previous state is repeated in the initial `if` condition and in each case of the switch statement. This can be simplified by always copying the last state from the `states` array inside the loop, and initializing `states` with a copy of the initial state.